keps/sig-node/4383-dra-semantic-model-mode-switching/README.md

```diff
@@ -0,0 +1,898 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+One or more "semantic model" need to be defined and implemented for "semantic
+parameters" ([KEP
+#4381](https://github.com/kubernetes/enhancements/issues/4381)) to be usable
+with Dynamic Resource Allocation (DRA).
+
+This KEP defines a semantic model that roughly corresponds in functionality to
+what the initial set of DRA drivers for GPUs supported with vendor CRDs.
+
+## Motivation
+
+### Goals
+
+* Enable DRA GPU drivers to support cluster autoscaling.
+
+### Non-Goals
+
+* Provide a model that can handle all future hardware. Other models may be
+  needed for things that don't fit into the schema defined here.
+
+## Proposal
+
+The resource model is that each node contains a certain set of resource
+instances. Each resource instance can be in one of several different, finite
+modes. Each mode has different attributes. Some of these attributes are labels
+that allow selecting a suitable instance, others are counters of things that
+are available in the mode for consumption by different resource claims.
+
+Each instance can be only in one mode at once. It can switch between modes only
+when it is not currently in use.
+
+### Risks and Mitigations
+
+Adding another semantic model doesn't add any additional risks. There is a
+slight performance impact when checking whether the builtin controller of the
+semantic model needs to do something for a claim or NodeResourceCapacity object,
+but that impact should be minimal when the model is not in use.
+
+This resource model is fairly complex. Finding the right instance and mode of
+that instance to satisfy a request has to be done by iterating over them, which
+can be expensive. To mitigate that, requests should be written so that "simple"
+checks are executed first to prune the search tree early.
+
+## Design Details
+
+The semantic model is hosted in
+`k/k/staging/src/k8s.io/dynamic-resource-allocation`:
+
+- `semantic/modeswitch`: the implementation of the builtin controller
+- `apis/modeswitch`: the versioned type(s)
+- `internal/apis/modeswitch`: the internal type for conversion and for use
+  by the implementation
+
+The only modification elsewhere is registering the builtin controller in the
+dynamic resource scheduler plugin if (and only if) the model is enabled.
+
+The semantic model uses a label selector for selecting instances and modes that
+is more capable than `metav1.LabelSelector`. For example, semantic version
```

Comment:
why not enhance the existing node selectors?

semantic version awareness seems like something universally useful.

Reply:
Because extending `metav1.LabelSelector` is modifying other existing GA APIs. That could be solved by being extremely careful about where and when to allow new operators, but is probably not something that should be attempted immediately when we don't even know whether this new selector proposal is useful.

Reply:
I'm not much in favor with the idea of having more than one way to express the same thing.

Is there anything else, other than semantic versioning, that this API allows that the existing node selectors doesn't support?
They already support OR (nodeSelectorTerms) and AND (matchExpressions).

Reply:
I had a similar concern about Karpenter - the EC2NodeClass API has “selectors” that are not Kubernetes selectors, they are actually AWS EC2 filters (which can match on `tags`, whereas Kubernetes doesn't have tags).

However, EC2NodeClass is vendor code and not part of the Karpenter core that the Kubernetes project has adopted. So, it's not nearly as much as a problem; we don't expect vendors to follow our guidelines, though it's nice when they do.

---

Adding a new comparison operator would OK by me (eg `SemVerGe` to compare to strings as semantic versions), because we already have some selectors that support numeric comparison and others that don't. I **don't** like the idea of adding `type` into selectors; right now, the operator infers a type (eg, `Lt` infers Quantity). We should stick with that.

Reply:
> Is there anything else, other than semantic versioning, that this API allows that the existing node selectors doesn't support?

Deeply nested expressions. With this grammar one can do `a == 1 || (b > 100 && (c != 42 || d == 1000) )`.

> I don't like the idea of adding type into selectors; right now, the operator infers a type (eg, Lt infers Quantity).

We need to implement `==` for strings, quantity, and semver. I can see us using `Eq" for semver, `==` for quantity, and "Is" for strings.

But what do we then use for '<' for strings?

None of this is an issue when defining the type separately.

Reply:
I like the idea of using a CEL expression stored inside a string.